zrml/swaps/src/lib.rs

```diff
@@ -770,6 +1341,182 @@ mod pallet {
+                                 provided than currently reserved.
+                            Pool: {:?}, User: {:?}, Unreserved: {:?}, Previously reserved: {:?}",
+                                pool_id,
+                                provider_address,
+                                transfered,
+                                subsidy
+                            );
+                        }
+
+                        T::Shares::deposit(pool_shares_id, &provider_address, transfered)?;
+                        total_balance.saturating_add(transfered);
+                    }
+
+                    ensure!(total_balance >= T::MinSubsidy::get(), Error::<T>::InsufficientSubsidy);
+                    pool.total_subsidy = Some(total_balance);
+
+                    // Assign the initial set of outstanding assets to the pool account.
+                    let outstanding_assets_per_event =
+                        T::RikiddoSigmoidFeeMarketEma::initial_outstanding_assets(
+                            pool_id,
+                            pool.assets.len().saturated_into::<u32>().saturating_sub(1),
+                            total_balance,
+                        )?;
+
+                    for asset in pool.assets.iter().filter(|e| **e != base_asset) {
+                        T::Shares::deposit(*asset, &pool_account, outstanding_assets_per_event)?;
+                    }
+
+                    pool.pool_status = PoolStatus::Active;
+                    Self::deposit_event(Event::SubsidyCollected(pool_id, total_balance));
+                    Ok(())
+                });
+
+                if let Err(err) = result {
+                    if err == Error::<T>::InsufficientSubsidy.into() {
+                        return Ok(ResultWithWeightInfo {
+                            result: false,
+                            weight: T::WeightInfo::end_subsidy_phase(
+                                total_assets.saturated_into(),
+                                total_providers.saturated_into(),
+                            ),
+                        });
+                    }
+
+                    return Err(err);
+                }
+
+                Ok(ResultWithWeightInfo {
+                    result: true,
+                    weight: T::WeightInfo::end_subsidy_phase(
+                        total_assets.saturated_into(),
+                        total_providers.saturated_into(),
+                    ),
+                })
+            };
+
+            with_transaction(|| {
+                let output = do_mutate();
+                match output {
+                    Ok(res) => {
+                        if res.result {
+                            TransactionOutcome::Commit(Ok(res))
+                        } else {
+                            TransactionOutcome::Rollback(Ok(res))
+                        }
+                    }
+                    Err(err) => TransactionOutcome::Rollback(Err(err)),
+                }
+            })
```

Comment:
on_initialize -> process_subsidy_collecting_markets -> end_subsidy_phase

`on_initialize` already calls `with_transaction`. IMHO, `with_transaction` should only be used on pallet calls or pallet hooks.

Reply:
The problem here is that the code has to introduce state changes to see if there are some inconsistencies, i.e. some other functions has unreserved more than it should and therefore funds reserved for subsidy are not sufficient anymore in total. In this case the function does not fail though - it just reverts all changes and returns that the operation was successful, but no state transition occured because not as much could be unreserved as was reserved / registered before.
I can't think of another solution rn and with the release pressing, I'd like to move this to the next iteration, since it does not affect functionality (it's not a bug, maybe just not the most elegant solution). Let's create an issue for that.

Reply:
An issue looks good.